src/main/scala/org/apache/spark/ml/feature/RapidsPCA.scala

```diff
@@ -168,26 +173,58 @@ class RapidsPCAModel(
    */
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema, logging = true)
+    val cols_A = dataset.select($(inputCol)).first.size
 
-    // TODO(rongou): make this faster and re-enable.
-    //    if (getUseGemm) {
-    //      val transformed = dataset.toDF().rdd.mapPartitions(iterator => {
-    //        val gpuID = TaskContext.get().resources()("gpu").addresses(0)
-    //        val partition = iterator.toList
-    //        val A = Matrices.fromVectors(partition.map(_.getAs[Vector]($(inputCol)))).toDense
-    //        val C = Matrices.zeros(partition.length, getK).toDense
-    //        CUBLAS.gemm_b(A, pc, C, gpuID)
-    //        C.rowIter.zip(partition.iterator).map { case (v, r) =>
-    //          Row.fromSeq(r.toSeq ++ Seq(v))
-    //        }
-    //      })
-    //      dataset.sparkSession.createDataFrame(transformed, outputSchema)
-    //    }
-    //    else {
-    val transposed = pc.transpose
-    val transformer = udf { vector: Vector => transposed.multiply(vector) }
-    dataset.withColumn($(outputCol), transformer(col($(inputCol))), outputSchema($(outputCol)).metadata)
-    //    }
+    /**
+     * UDF class to speedup transform process of PCA
+     */
+    class gpuTransform extends Function[Array[Double], Array[Double]] with RapidsUDF with Serializable {
+      override def evaluateColumnar(args: ColumnVector*): ColumnVector = {
+        require(args.length == 1, s"Unexpected argument count: ${args.length}")
+        val input = args.head
+        val rows_A = input.getRowCount.toInt
+        val AdevAddr = input.getData.getAddress
+        var C: Long = 0
+        // A: raw data, B: pc, C: output, deviceID= 0 for test
+        C = RAPIDSML.gemm_test(RAPIDSML.CublasOperationT.CUBLAS_OP_N.id,
+          RAPIDSML.CublasOperationT.CUBLAS_OP_N.id,
+          rows_A, pc.numCols, cols_A, 1.0, AdevAddr, cols_A, pc, cols_A, 0.0, C, rows_A, 0)
+        val dmb = buildDeviceMemoryBuffer(C, (rows_A * pc.numCols * DType.FLOAT64.getSizeInBytes).toLong)
+        val childColumn = new ColumnVector(DType.FLOAT64, rows_A.toLong, Optional.of(0), dmb,
+          null, null)
+        val offsetCV = ColumnVector.sequence(Scalar.fromInt(0), Scalar.fromInt(pc.numCols), rows_A)
+        val toClose = new java.util.ArrayList[DeviceMemoryBuffer]()
+        toClose.add(dmb)
```

Comment:
dmb will be taken by childColumn, no need to be closed

Reply:
removed from toClose